external/sql/storm-sql-core/src/jvm/org/apache/storm/sql/compiler/backends/trident/TridentLogicalPlanCompiler.java

```diff
@@ -0,0 +1,344 @@
+                    .project(new Fields(temporaryOutputFieldNames))
+                    .each(new Fields(temporaryOutputFieldNames), new ForwardFunction(), new Fields(outputFieldNames))
+                    .project(new Fields(outputFieldNames))
+                    .name(stageName);
+        }
+    }
+
+    @Override
+    public IAggregatableStream visitFilter(Filter filter, List<IAggregatableStream> inputStreams) throws Exception {
+        if (inputStreams.size() > 1) {
+            throw new RuntimeException("Filter is a SingleRel");
+        }
+
+        Stream inputStream = inputStreams.get(0).toStream();
+        String stageName = getStageName(filter);
+
+        try (StringWriter sw = new StringWriter(); PrintWriter pw = new PrintWriter(sw)) {
+            pw.write("import org.apache.storm.tuple.Values;\n");
+
+            ExprCompiler compiler = new ExprCompiler(pw, typeFactory);
+            String ret = filter.getCondition().accept(compiler);
+            pw.write(String.format("\nreturn %s;", ret));
+            final String expression = sw.toString();
+
+            return inputStream.filter(new EvaluationFilter(expression)).name(stageName);
+        }
+    }
+
+    @Override
+    public IAggregatableStream visitAggregate(Aggregate aggregate, List<IAggregatableStream> inputStreams) throws Exception {
+        if (inputStreams.size() > 1) {
+            throw new RuntimeException("Aggregate is a SingleRel");
+        }
+
+        Stream inputStream = inputStreams.get(0).toStream();
+        String stageName = getStageName(aggregate);
+
+        List<String> groupByFieldNames = new ArrayList<>();
+        for (Integer idx : aggregate.getGroupSet()) {
+            String fieldName = inputStream.getOutputFields().get(idx);
+            groupByFieldNames.add(fieldName);
+        }
+
+        Fields groupByFields = new Fields(groupByFieldNames);
+        GroupedStream groupedStream = inputStream.groupBy(groupByFields);
+
+        List<Stream> joiningStreams = new ArrayList<>();
+        List<Fields> joiningFields = new ArrayList<>();
+        List<String> newOutputFields = new ArrayList<>();
+        newOutputFields.addAll(groupByFieldNames);
+
+        for (AggregateCall call : aggregate.getAggCallList()) {
+            Stream aggregateStream = handleAggregateCall(groupedStream, groupByFieldNames, inputStream, call);
+
+            Fields outputFields = aggregateStream.getOutputFields();
+            // assuming handleAggregateCall does project so that new output field is always placed to end of fields
+            String newFieldName = outputFields.get(outputFields.size() - 1);
+
+            joiningStreams.add(aggregateStream);
+            joiningFields.add(groupByFields);
+            newOutputFields.add(newFieldName);
+        }
+
+        return topology.join(joiningStreams, joiningFields, new Fields(newOutputFields), JoinType.INNER).name(stageName);
```

Comment:
http://storm.apache.org/releases/1.0.1/Trident-API-Overview.html

> The benefits of CombinerAggregators are seen when you use them with the aggregate method instead of partitionAggregate. In that case, Trident automatically optimizes the computation by doing partial aggregations before transferring tuples over the network.

While reading same tuples for multiple time feels redundant, handling all aggregations in one base function will drop the chance for optimization like this.

I'd like to try out Trident optimizations for all of cases, and address optimization for Trident itself when needed.

Btw, I also would like to replace this to chainedAgg(). I'll address this.